### DISCUSSION TITLE
Use entity Manager inside entity

### DISCUSSION THREAD

**Discussion by azhar2202** [symfony/symfony#50383] [Q&A]
i have problem about use entity manager inside entity, i use entity manager because i need to call my helper that have parameter entity manager and i use static function in my helper

     use Doctrine\ORM\EntityManagerInterface;
     private ?EntityManagerInterface $entityManager;

    public function __construct(EntityManagerInterface $entityManager)
    {
        $this->id = Uuid::v4();
        $this->userTwoFactors = new ArrayCollection();
        $this->role = new ArrayCollection();
        $this->ownedGroups = new ArrayCollection();
        $this->groupMembers = new ArrayCollection();
        $this->entityManager = $entityManager;
    public function getRoles(): array
    {
   
        $plainRoles = $this->getDirectRoles();

      
        if (null !== $this->getPegawai()) {
            $arrayOfRoles = [];

            // If on leave, assign ROLE_ON_LEAVE
            if ($this->getPegawai()->getOnLeave()) {
                $arrayOfRoles[] = ['ROLE_ON_LEAVE'];
            }

            // make sure that retired person doesn't get role
            if (!$this->getPegawai()->getPensiun()) {
                /** @var JabatanPegawai $jabatanPegawai */
                foreach ($this->getPegawai()->getJabatanPegawais() as $jabatanPegawai) {
                    // Only process active jabatans
                    if ($jabatanPegawai->getTanggalMulai() <= new DateTimeImmutable('now')
                        && ($jabatanPegawai->getTanggalSelesai() >= new DateTimeImmutable('now')
                            || null === $jabatanPegawai->getTanggalSelesai())
                    ) {
                        $arrayOfRoles[] = array_values(
                            RoleHelper::getPlainRolesNameFromJabatanPegawai($this->entityManager ,$jabatanPegawai)
                        );
                    }
                }
            } else {
                return ['ROLE_RETIRED'];
            }

            $plainRoles = array_values(array_merge($plainRoles, ...$arrayOfRoles));
        }

        return array_values(array_unique($plainRoles));
    }

i try to use EntityManager inside the __construct and try to use it in function getRoles(), but this return error

"Typed property App\Entity\User\User::$entityManager must not be accessed before initialization"

this my helper

        public static function getRolesFromJabatanPegawaiCustom(EntityManagerInterface $entityManager, JabatanPegawai $jabatanPegawai): array
    {
      
        $roles = [];
        $jabatan = $jabatanPegawai->getJabatan();
        $unit = $jabatanPegawai->getUnit();
        $kantor = $jabatanPegawai->getKantor();
        $jenisKantorKantor = $kantor?->getJenisKantor();
        $jenisKantorUnit = $unit?->getJenisKantor();
        $pegawai = $jabatanPegawai->getPegawai();
        $pegawaiId = $pegawai->getId();

        $role = $jabatan->getRoles();

        if ($jenisKantorKantor === $jenisKantorUnit
            && null !== $jenisKantorKantor
            && null !== $jenisKantorUnit
        ) {
            $jenisKantor = $jenisKantorKantor;
        } else {
            $jenisKantor = null;
        }

 
        if (null !== $jabatan) {
            // direct role from jabatan/ jabatan unit/ jabatan kantor/ combination
            foreach ($jabatan->getRoles() as $role) {
                 if(false === $role->isOperator()){
                    if (2 === $role->getJenis()) {
                        $roles[] = $role;
                    } elseif (8 === $role->getJenis() && $role->getUnits()->contains($unit)) {
                        $roles[] = $role;
                    } elseif (9 === $role->getJenis() && $role->getKantors()->contains($kantor)) {
                        $roles[] = $role;
                    } elseif (10 === $role->getJenis()
                        && $role->getUnits()->contains($unit)
                        && $role->getKantors()->contains($kantor)
                    ) {
                        $roles[] = $role;
                    } elseif (11 === $role->getJenis()
                        && $role->getUnits()->contains($unit)
                        && $role->getJenisKantors()->contains($jenisKantor)
                    ) {
                        $roles[] = $role;
                    }
                 }elseif (true === $role->isOperator()){
                     $roleCombination = $entityManager
                         ->getRepository(JabatanPegawai::class)
                         ->findRoleCombinationByPegawai($pegawaiId);
                     if(2 === $role->getJenis() && $role->getNama()) {
                         $roles[] = $roleCombination;
                     }
                 }
            }

            $eselon = $jabatan->getEselon();
            if (null !== $eselon) {
                foreach ($eselon->getRoles() as $role) {
                    if (5 === $role->getJenis()) {
                        $roles[] = $role;
                    }
                }
            }
        }

        if (null !== $unit) {
            foreach ($unit->getRoles() as $role) {
                if (3 === $role->getJenis()) {
                    $roles[] = $role;
                }
            }
        }

        if (null !== $kantor) {
            foreach ($kantor->getRoles() as $role) {
                if (4 === $role->getJenis()) {
                    $roles[] = $role;
                }
            }
        }

        if (null !== $jabatan->getEselon()) {
            foreach ($jabatan->getEselon()->getRoles() as $role) {
                if (5 === $role->getJenis()) {
                    $roles[] = $role;
                }
            }
        }

        if (null !== $jenisKantor) {
            foreach ($jenisKantor->getRoles() as $role) {
                if (6 === $role->getJenis()) {
                    $roles[] = $role;
                }
            }
        }

        return $roles;
    }

    /**
     * @param JabatanPegawai $jabatanPegawai
     * @param EntityManagerInterface $entityManager
     * @return array
     */
    public static function getPlainRolesNameFromJabatanPegawai(EntityManagerInterface $entityManager, JabatanPegawai $jabatanPegawai): array
    {
        $roles = self::getRolesFromJabatanPegawaiCustom($entityManager, $jabatanPegawai);
        $plainRoles = [];

        /** @var Role $role */
        foreach ($roles as $role) {
            if ($role->getStartDate() <= new DateTimeImmutable('now')
                && ($role->getEndDate() >= new DateTimeImmutable('now')
                    || null === $role->getEndDate())
            ) {
                $plainRoles[] = $role->getNama();
            }
        }

        return array_values(array_unique($plainRoles));
    }

what is the best practice to solve my problem? thank you

**Comment by MatTheCat** [comment]
Entities shouldn’t access their entity manager when using Doctrine. Implement your `getRoles` method in a [repository](https://symfony.com/doc/current/doctrine.html#querying-for-objects-the-repository) (or dedicated service) instead.

**Reply by azhar2202** [reply]
can you explain or give me example how to make dedicated service? thank you

**Reply by MatTheCat** [reply]
See https://symfony.com/doc/current/service_container.html#creating-configuring-services-in-the-container
